### DISCUSSION TITLE
Style spec testing and development features

### DISCUSSION THREAD

**Discussion by davidemerritt** [maplibre/maplibre-style-spec#553] [Ideas]
Hello 👋 ,

I am on the newish side to mapbox style-spec and GIS rendering generally so I may be missing some obvious tooling or approaches here. 

I have found that it is hard to test, develop and debug complex expressions within the maplibre style spec.  This is especially true for nested expressions with multiple cases, coalesce, match etc where the feature attributes can be quite varied. 

Is there a tool, cli, web, etc that helps to test and evaluate the style spec expressions against a set of known attributes? How is this general development problem approached currently? 

Appreciate any responses!

**Comment by HarelM** [comment]
In term of style editor, maputnik is probably the goto from our perspective as you get immediate feedback.
You can use some of the unit tests here (expression test) to make sure an expression is valid.
Other than that, you might be able to use the typescript definitions, although I think they are not fully correct, but the more people use them and improve them, the better we will be.

Generally speaking, since this is a language, this kind of tooling is needed, but except the above I don't think there's something good.

Any help in that area would be welcome.

I would also consider doing the heavy lifting in the tile generation, making the expressions as simple as possible.

**Reply by davidemerritt** [reply]
thanks I'll check it out, I was hoping for a utility that lets you run `evaluateFeature(feature, style) => output` but that may be optimistic :)
